scylla/tests/integration/caching_session.rs

```diff
@@ -0,0 +1,140 @@
+                .await
+                .unwrap();
+            // Assumption: all nodes have the same number of shards
+            let nr_shards = match session.get_cluster_state().get_nodes_info().first() {
+                Some(node_info) => match node_info.sharder() {
+                    Some(sharder) => sharder.nr_shards,
+                    None => NonZero::new(1).unwrap(), // If there is no sharder, assume 1 shard
+                },
+                None => panic!("No nodes information available"),
+            };
+
+            // Consume all feedbacks so far to ensure we will not count something unrelated.
+            let _feedbacks: usize = feedback_rxs.iter_mut().map(consume_current_feedbacks).sum();
+
+            let caching_session: CachingSession = CachingSession::from(session, 100);
+            let insert_1_b_c = "INSERT INTO tab (a, b, c) VALUES (1, ?, ?)";
+            let insert_2_b_c = "INSERT INTO tab (a, b, c) VALUES (2, ?, ?)";
+            let insert_3_b_c = "INSERT INTO tab (a, b, c) VALUES (3, ?, ?)";
+            let insert_4_b_c = "INSERT INTO tab (a, b, c) VALUES (4, ?, ?)";
+            let mut batch = Batch::new(BatchType::Logged);
+            batch.append_statement(insert_1_b_c);
+            batch.append_statement(insert_2_b_c);
+            batch.append_statement(insert_3_b_c);
+            batch.append_statement(insert_4_b_c);
+            // First batch that should generate prepares for each shard.
+            let batch_values = vec![(1, 1), (2, 2), (3, 3), (4, 4)];
+            caching_session.batch(&batch, batch_values.clone()).await.unwrap();
+            // Few extra runs. Those batches should not result in any prepares being sent.
+            caching_session.batch(&batch, batch_values.clone()).await.unwrap();
+            caching_session.batch(&batch, batch_values.clone()).await.unwrap();
+            caching_session.batch(&batch, batch_values.clone()).await.unwrap();
+            caching_session.batch(&batch, batch_values.clone()).await.unwrap();
+
+            let prepared_batch_res_rows: Vec<(i32, i32, i32)> = caching_session
+                .execute_unpaged("SELECT * FROM tab", &[])
+                .await
+                .unwrap()
+                .into_rows_result()
+                .unwrap()
+                .rows()
+                .unwrap()
+                .map(|r| r.unwrap())
+                .collect();
+
+
+            // Verify the data from inserts
+            let mut prepared_batch_res_rows = prepared_batch_res_rows;
+            prepared_batch_res_rows.sort();
+            let expected_rows = vec![(1, 1, 1), (2, 2, 2), (3, 3, 3), (4, 4, 4)];
+            assert_eq!(prepared_batch_res_rows, expected_rows);
+
+
+            let feedbacks: usize = feedback_rxs.iter_mut().map(consume_current_feedbacks).sum();
+
+            let expected_feedbacks = (4 + 1) * nr_shards.get() as usize * 3; // 4 inserts + 1 select, 3 nodes
+            assert_eq!(feedbacks, expected_feedbacks);
+
```

Comment:
Can you verify the feedbacks amount after first execution of the batch, and then again after other executions to verify there are no new prepares?

Reply:
I've added more frequent checks. Now the number of prepares is verified after each query. I've also put those numbers into descriptive variables like `batch_size`, `cluster_size`.